### DISCUSSION TITLE
Does the windows environment cause multiple client links to block

### DISCUSSION THREAD

**Discussion by zhenzi0322** [miguelgrinberg/Flask-SocketIO#1852] [Q&A]
Hello, I am using Flask-SocketIO to process tasks, if a user sends multiple tasks, it will cause the link Flask-SocketIO to block, my environment is windows

**Comment by miguelgrinberg** [comment]
Flask-SocketIO does not block. My guess is that the tasks that you are running are blocking your server. Or if you are using an asynchronous server (eventlet, gevent) then you are writing blocking tasks that are incompatible with these frameworks.

May I ask you to not write issues on this package when you have a question? There is a discussions board for questions. Thank you.

**Reply by zhenzi0322** [reply]
Ok, I'm using start_background_task to process the task in the background using a concurrent thread, will this affect the blocking?

**Reply by miguelgrinberg** [reply]
What you do in your task is what will may cause blocking. The `start_background_task()` always starts background tasks in a way that is compatible with the async framework that you are using.

**Reply by zhenzi0322** [reply]
> What you do in your task is what will may cause blocking. The `start_background_task()` always starts background tasks in a way that is compatible with the async framework that you are using.

Since there are tasks executing in the background, it blocks the Flask-SocketIO link. Is there any way to make my task not block the Flask-SocketIO link?
Is it possible to use the following startup method to not block the Flask-SocketIO link.

```shell
gunicorn -k geventwebsocket.gunicorn.workers.GeventWebSocketWorker -w 1 app:app
```

Examples：
For example: I open a client A to link Flask-SocketIO, client A sends a hundred tasks to start_background_task for processing, and then I open a new client B. At this point client B blocks, causing Flask-SocketIO to wait for the link.

**Reply by miguelgrinberg** [reply]
@zhenzi0322 I think you need to familiarize yourself with gevent and learn how concurrency works under this framework. Background tasks should be perfectly fine, as long as they don't block the gevent loop. If you are not familiar with asynchronous development, then my recommendation is that you use the `threading` concurrency mode, which is based on regular threads and will be easier to manage for you.

**Reply by zhenzi0322** [reply]
> @zhenzi0322 I think you need to familiarize yourself with gevent and learn how concurrency works under this framework. Background tasks should be perfectly fine, as long as they don't block the gevent loop. If you are not familiar with asynchronous development, then my recommendation is that you use the `threading` concurrency mode, which is based on regular threads and will be easier to manage for you.

Ok, Because I send a message to the foreground via emit after processing a task in the background task. And that's when it causes blocking. The background task itself does not block if it does not send an emit.


Example:
```python
from gevent import monkey

monkey.patch_all()
from gevent.pool import Pool
import random
import time
from config import socketio


def background_thread():
    """start_background_task"""
    p = Pool(size=20)
    count = 0
    while True:
        count += 1
        value = queue.get_wait()
        if value:
            time.sleep(random.randint(2, 8))
            p.spawn(socketio.emit, "result", {"data": "ok"}, to="xxx")
```

**Reply by miguelgrinberg** [reply]
I don't understand what the problem is. The emit function is non-blocking, all it does is store the message in a queue and returns. The actual data transfer happens in the background. So I don't see how emit can be blocking. There is also no need to put the emit call in a pool, you can just call `socket.emit()` directly in the function and you'll get the same result.
